### PR TITLE
#331 etags on uploads

### PR DESCRIPTION
**Pull request by julianseeger**
resolves Qabel/qabel-desktop#331

### CODE REVIEW

File: box/src/main/java/de/qabel/box/storage/hash/Hasher.kt [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package de.qabel.box.storage.hash
+
+import java.io.IOException
+import java.nio.file.Path
+
+interface Hasher {
+    @Throws(IOException::class)
+    fun getMHash(file: Path): String
```
Comment (enkore) [inline box/src/main/java/de/qabel/box/storage/hash/Hasher.kt]: get*M*Hash?

File: box/src/main/java/de/qabel/box/http/HttpWriteBackend.kt [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+            httpPost.entity = InputStreamEntity(content)
+
+            httpclient.execute(httpPost).use { response ->
+                val status = response.statusLine.statusCode
+                if (status == 404 || status == 403) {
+                    throw QblStorageNotFound("File not found")
+                }
+                if (status >= 300) {
+                    throw QblStorageException("Upload error")
+                }
```
Comment (enkore) [inline box/src/main/java/de/qabel/box/http/HttpWriteBackend.kt]: while you're touching this code, can you improve the exception here bit? it should contain the error code and maybe the response from the server.

File: box/src/main/java/de/qabel/box/storage/hash/Hasher.kt (hunk 1 above)
Comment (julianseeger) [inline box/src/main/java/de/qabel/box/storage/hash/Hasher.kt]: the `mhash` predates this PR: https://github.com/Qabel/qabel-core/search?utf8=%E2%9C%93&q=mhash
It was an evolution from mtime to "still mtime but let's not call it time" ;)
But sure, this Interface doesn't need to know anything about that legacy

File: box/src/test/java/de/qabel/box/http/HttpWriteBackendTest.kt [hunk 3]
```diff
@@ -12,14 +16,30 @@ class HttpWriteBackendTest {
...
-        //writeBackend.uploadIfOld("name", )
-        TODO()
+        writeBackend.upload("file", fakeDataStream())
+        writeBackend.upload("file", fakeDataStream(), "not the current etag")
+    }
+
+    @Test
+    fun uploadsSuccessfulIfEtagIsUpToDate() {
+        val upload = writeBackend.upload("file", fakeDataStream())
+        writeBackend.upload("file", fakeDataStream("new"), upload.etag)
```
Comment (audax) [inline box/src/test/java/de/qabel/box/http/HttpWriteBackendTest.kt]: Assert that an etag actually exists and is not the empty string

Comment (julianseeger) [inline box/src/test/java/de/qabel/box/http/HttpWriteBackendTest.kt]: done

File: box/src/main/java/de/qabel/box/http/AbstractHttpStorageBackend.kt [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    /**
+     * This method can be overwritten to modify the request before it is executed.
+     * You may want to add special headers here.
+     */
+    protected open fun prepareRequest(request: HttpRequest) {
+
+    }
+
+    companion object {
+        // Number of http connections to S3
```
Comment (audax) [inline box/src/main/java/de/qabel/box/http/AbstractHttpStorageBackend.kt]: not S3 anymore

File: box/src/main/java/de/qabel/box/http/HttpReadBackend.java [hunk 5]
```diff
@@ -32,16 +32,16 @@ public StorageDownload download(String name) throws QblStorageException {
 
     @Override
     public StorageDownload download(String name, String ifModifiedVersion) throws QblStorageException, UnmodifiedException {
-        logger.info("Downloading " + name);
-        URI uri = root.resolve(name);
+        info("Downloading " + name);
+        URI uri = getRoot().resolve(name);
```
Comment (audax) [inline box/src/main/java/de/qabel/box/http/HttpReadBackend.java]: Why is it getRoot() suddenly?

**Comment by audax** [comment]
You may resolve the remaining "issues" in a seperate pr ;)

File: box/src/main/java/de/qabel/box/http/AbstractHttpStorageBackend.kt [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+        // The default was too low, 20 works. Further testing may be required
+        // to find the best amount of connections.
+        protected val CONNECTIONS = 50
```
Comment (enkore) [inline box/src/main/java/de/qabel/box/http/AbstractHttpStorageBackend.kt]: using pipelining should reduce that number to essentially 1. create ticket?

**Comment by audax** [comment]
Create ticket.

File: box/src/main/java/de/qabel/box/http/HttpWriteBackend.kt [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+import de.qabel.box.storage.StorageWriteBackend
+import de.qabel.box.storage.exceptions.QblStorageException
+import de.qabel.box.storage.exceptions.QblStorageNotFound
+import de.qabel.core.logging.QabelLog
+import org.apache.http.client.methods.CloseableHttpResponse
+import org.apache.http.client.methods.HttpDelete
+import org.apache.http.client.methods.HttpPost
+import org.apache.http.client.utils.DateUtils
+import org.apache.http.entity.InputStreamEntity
+import sun.plugin.dom.exception.InvalidStateException
```
Comment (audax) [inline box/src/main/java/de/qabel/box/http/HttpWriteBackend.kt]: wtf

Comment (enkore) [inline box/src/main/java/de/qabel/box/http/HttpWriteBackend.kt]: Should be IllegalStateException?

**Pull request merged**